WIP : IGNORE - running tests on TC

bobvawter opened the pull request.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/31171)
<!-- Reviewable:end -->

bobvawter committed:
sql: Align timestamp math with PostgreSQL

Our implementation of time+duration math delegates to golang's
time.Time.AddDate(), which has different normalization semantics when compared
to PostgreSQL when normalizing invalid month-of-day values.

The most likely input case is that the duration will have no month component
and we can fall back to go's implementation.  Another fast-path of checking the
day-of-month allows us to skip the special-case logic for start dates that
aren't the 29th, 30th, or 31st of a month.

```
name                                      old time/op    new time/op    delta
Add/fast-path-by-no-months-in-duration-8    63.2ns ± 4%    64.8ns ± 1%      ~     (p=0.151 n=5+5)
Add/fast-path-by-day-number-8               70.5ns ± 5%    87.5ns ± 0%   +24.12%  (p=0.008 n=5+5)
Add/no-adjustment-8                         70.6ns ± 3%   196.4ns ± 0%  +178.11%  (p=0.008 n=5+5)
Add/with-adjustment-8                       68.1ns ± 3%   218.2ns ± 1%  +220.60%  (p=0.008 n=5+5)

(No difference in allocations)
```

Resolves #30976

Release note (sql change): The normalized results of certain timestamp +
duration operations involving year or month durations have been adjusted to
agree with the values returned by PostgreSQL.  In cases such as
'2018-01-31'::TIMESTAMP + '1 month', where an intermediate result of February
31st needs to be normalized, previous versions of CockroachDB would "round up"
to March 1.  Instead, we would now "round down" to February 28th to agree with
the values returned by PostgreSQL.  This change also affects the results of
`generate_sequence()` when used with timestamps.

Users can control this behavior on a per-session or on a cluster-wide basis.
The session variable `duration_addition_mode` or cluster setting
`sql.defaults.duration_addition_mode` can be set to `legacy` for the old
behavior.

Release note: None
